Fixes #22733 - Add customizable columns to Red Hat Subscriptions Page

chalettu opened the pull request:
Demo of functionality. 
![subscriptions_page_demo](https://user-images.githubusercontent.com/230549/41424690-05fe4a6e-6fbc-11e8-9ae7-36aceafdc311.gif)

ehelms commented:
@chalettu I sent you an org invite, please set yourself public if you accept. Testing will happen automatically for you from there on out.

ehelms commented:
Is it possible to make the column change react to clicking or un-clicking? I find the fact you have to click off of the dropdown obvious as a confirmation mechanism.

waldenraines commented:
> Is it possible to make the column change react to clicking or un-clicking? I find the fact you have to click off of the dropdown obvious as a confirmation mechanism.

I don't understand this.  Are you saying to change the columns with each new selection of a column from the list of columns?  I.e. if I click End Date in the dropdown that column immediately appears?  

If so, the only problem I see with that is the race condition of the user clicking quickly enough to select a new column before the API call completes for the previous one the selected.  This could cause odd behavior.

chalettu commented:
So I do have a code solution to this I am about to submit.  @waldenraines , I am still setting it to where redux doesn't save the columns until you close the tooltip but I am setting it up to change the columns as you change them.  It should solve both issues.

chalettu commented:
Here is a updated demo.  The page now reacts when a column is clicked or unchecked. 
![updated_demo](https://user-images.githubusercontent.com/230549/41432026-03dd148e-6fd2-11e8-9f8c-1ffc3fcac968.gif)

waldenraines commented on webpack/move_to_pf/OptionTooltip/index.js:
```diff
@@ -0,0 +1,77 @@
+  }
+  renderTooltip() {
+    return (
+      <Popover id="optionTooltip" className="option-tooltip">
```
Is this ID used anywhere?

waldenraines commented on webpack/move_to_pf/OptionTooltip/index.js:
```diff
@@ -0,0 +1,77 @@
+    const { icon } = this.props;
+    const rootClose = true;
+    const onOpen = () => {
+      this.setState({ toolTipOpen: true });
```
Nitpick: we mix toolTip and tooltip in this file.

waldenraines commented on webpack/redux/actions/Settings/tables.js:
```diff
@@ -0,0 +1,85 @@
+import { foremanApi, userId } from '../../../services/api';
```
I would argue that this file should live in `webpack/scenes/Settings` as it's similar to `webpack/scenes/Organizations`.  The files in the redux directory will be moved in the future.

chalettu replied:
I went ahead and moved all redux work for tables to the recommended folder you mentioned above.

waldenraines commented on webpack/redux/actions/Settings/tables.js:
```diff
@@ -0,0 +1,85 @@
+    .catch((result) => {
+      const { response } = result;
+      const errorMessage = `Resource table_preference not found by id '${table}'`;
+      if (response.status === 404 && response.data.error.message === errorMessage) {
```
Is there a case where a 404 would happen where the error messages don't equal?

chalettu replied:
I removed this altogether from the PR because I changed it to where our loadTables action loads all tables and got rid of the notion of retrieving individual tables.

waldenraines commented on webpack/scenes/Subscriptions/Subscriptions.scss:
```diff
@@ -12,3 +12,10 @@
+  margin-left: 10px;
+  margin-top: 3px;
+  width: 10px;
+}
```
Should this be in the `OptionTooltip.scss` file?

chalettu replied:
I thought about that and decided against it because this pertained to where the tooltip is positioned on the subscriptions page and not necessary where you might want it positioned on anywhere else it might be used within the application.   I tried to keep the scss for the tooltip a little more generic so it wouldn't be hard to migrate it to PF in the future.

waldenraines commented:
Also you'll want to combine your commits before we can merge this.

ehelms commented:
@chalettu Perfect.

Does the CSV export, only export the selected visible columns?

waldenraines commented:
> Does the CSV export, only export the selected visible columns?

No, it exports all columns.  We could probably make some change to the API to support this behavior if desired.

chalettu commented on webpack/move_to_pf/OptionTooltip/index.js:
```diff
@@ -0,0 +1,77 @@
+    super(props);
+
+    this.state = {
+      options: this.props.options,
```
Thank you for the review and link to the correct pattern.  I updated code based on your feedback.

chalettu commented on webpack/redux/reducers/Settings/tables.js:
```diff
@@ -0,0 +1,42 @@
+    case UPDATE_TABLE_COLUMNS_SUCCESS:
+      return Immutable({
+        loading: false,
+        ...action.response,
```
Good catch!  I had originally intended to build it this way but ended up changing it and forgot to revisit schema design.  I went ahead and rebuilt the redux data structures to rely on this pattern for design.

chalettu commented on webpack/scenes/Subscriptions/SubscriptionsPage.js:
```diff
@@ -23,13 +24,64 @@ import {
 class SubscriptionsPage extends Component {
   constructor(props) {
     super(props);
+    const options = [
```
@waldenraines , I can make this change but wanted a second opinion on whether you think this is a direction we should head in for how we should approach compiling a list of available columns.

waldenraines replied:
@sharvit any thoughts on this?  Will this be affected at all by the refactoring?

tstrachota replied:
@chalettu you're right, I forgot that we currently inject the controllers into the table schema which complicates the situation. @waldenraines if we refactor the controllers into redux actions then there will be bo need for controller injection I guess.

My main motivation for creating the column menu from table schema was to prepare tables for extendability from plugins. But it's a future feature. I'm ok to revisit this later and merge the current design. There's nothing wrong with it.

waldenraines commented:
[test]

waldenraines commented:
[test katello]
[test prprocessor]

chalettu commented:
[test prprocessor]

waldenraines commented:
@chalettu sorry, I caused some merge conflicts for you by merging #7467, please rebase.

ohadlevy commented:
@chalettu looks like a rebase is required, @waldenraines whats the status of this pr? thanks

waldenraines commented:
>  @waldenraines whats the status of this pr? thanks

It works well but I would like someone else to take another look at the code just to make sure I didn't miss anything (@amirfefer @tstrachota @sharvit :smile:)

tstrachota commented on webpack/move_to_pf/OptionTooltip/index.js:
```diff
@@ -0,0 +1,76 @@
+
+OptionTooltip.propTypes = {
+  icon: PropTypes.string.isRequired,
+  options: PropTypes.arrayOf(PropTypes.object).isRequired,
```
Nitpick: can you describe the shape here instead of just any object?

chalettu replied:
Updated shape of expected options

tstrachota commented on webpack/scenes/Settings/Tables/TableActions.js:
```diff
@@ -0,0 +1,73 @@
+
+export const loadTables = () => (dispatch) => {
+  dispatch({ type: TABLES_REQUEST, params: {} });
+  // foremanApi.delete(`/users/${userId}/table_preferences/Katello::Subscriptions`, {});
```
Some leftover?

chalettu replied:
I updated the branch and removed extra comment

tstrachota commented on webpack/scenes/Settings/Tables/__tests__/Table.fixtures.js:
```diff
@@ -0,0 +1,116 @@
+    type: TABLES_REQUEST,
+  },
+  {
+    error: null,
```
Could you test it with some real error so that we're sure the error message propagates into the action?

chalettu replied:
Updated with some real errors for testing

tstrachota commented on webpack/scenes/Settings/Tables/__tests__/Table.fixtures.js:
```diff
@@ -0,0 +1,116 @@
+    params: { name: 'Test', columns: [] },
+  },
+  {
+    error: undefined,
```
Ditto real data.

tstrachota commented on webpack/services/api/index.js:
```diff
@@ -103,3 +103,5 @@ export const foremanTasksApi = new ForemanTasksApi();
 // This node does not exist while testing
 export const orgId = orgNode ? orgNode.dataset.id : '1';
+const userNode = document.getElementById('user-id');
+export const userId = userNode ? userNode.dataset.id : '1';
```
I know you just copied it from the line above, but is this code safe? Shouldn't we rather mock this in tests and raise error when `orgNode` or `userNode` are missing to avoid unexpected behavior in production?
@waldenraines, @sharvit what do you think?

tstrachota commented:
@chalettu ping, this PR needs rebasing

chalettu commented:
@tstrachota , I went ahead and rebased this PR.  I also took @sharvit comments and worked on moving the state that I was using to use redux for the subscriptions page.   Please take a look and let me know if we have any more feedback or if it is good to merge.  @waldenraines can you take another look ?

chalettu commented on webpack/scenes/Subscriptions/SubscriptionsPage.js:
```diff
@@ -283,6 +386,10 @@ SubscriptionsPage.propTypes = {
+  loadTables: PropTypes.func.isRequired,
+  createColumns: PropTypes.func.isRequired,
+  updateColumns: PropTypes.func.isRequired,
+  settings: PropTypes.shape({}).isRequired,
```
In this case settings is controlled by redux.  One issue is that the table in question the user wants might possibly not exist.  It is a possible scenario so I figured it made sense to pass in the part of the redux state that controls settings and let the component figure out if the table existed or not.

sharvit replied:
Couldn't you just pass the `settings.tables[subscriptionTable]` to a prop called `subscriptionTableSettings`?
And your logic could be `if (subscriptionTableSettings) { ... } else { ... }`

chalettu commented:
@sharvit .  I moved the code I had that changed state internally to the subscription page out to redux that is used by the subscriptions page.  I rewrote the tests for this and it should all be working as expected.

sharvit commented on webpack/scenes/Settings/Tables/TableActions.js:
```diff
@@ -0,0 +1,73 @@
+      const { response } = result;
+      dispatch({
+        type: TABLES_FAILURE,
+        error: getResponseError(response),
```
Could you use `payload` instead of `error`?
IMO we should only dispatch type and payload as a formal API for inner communication.

sharvit commented on webpack/scenes/Settings/Tables/TableReducer.js:
```diff
@@ -0,0 +1,46 @@
+      return state.set('loading', true);
+
+    case TABLES_SUCCESS:
+      return Immutable({
```
Why do you create a new `Immutable` instead using `state.merge`?

sharvit commented on webpack/scenes/Settings/Tables/__tests__/TableActions.test.js:
```diff
@@ -0,0 +1,84 @@
+  mockReset();
+});
+
+describe('table actions', () => {
```
You could write much simpler test here if you would use the test-helper `testActionSnapshotWithFixtures`.
https://github.com/Katello/katello/blob/master/webpack/move_to_pf/test-utils/testHelpers.js#L58

We still don't have an example in `katello` but you can see an example in `foreman`:
https://github.com/theforeman/foreman/blob/fe74cc88d888cbf51f441c253e027b007440e77f/webpack/assets/javascripts/react_app/components/BreadcrumbBar/__tests__/BreadcrumbBarActions.test.js

If you want to, it could be nice to use the `testActionSnapshotWithFixtures`.

sharvit replied:
And you would be able to remove bunch of fixtures because it will create snapshots instead.

sharvit commented on webpack/move_to_pf/OptionTooltip/__tests__/OptionTooltip.test.js:
```diff
@@ -0,0 +1,38 @@
+import toJson from 'enzyme-to-json';
+import OptionTooltip from '../';
+
+describe('OptionTooltip', () => {
```
You could wrote much simpler tests if you would use the test-helper `testComponentSnapshotsWithFixtures`:
https://github.com/Katello/katello/blob/master/webpack/move_to_pf/test-utils/testHelpers.js#L31

Example using it:
https://github.com/Katello/katello/blob/fd8c2c012d62055b052a08820b37c76f35fef44e/webpack/move_to_foreman/components/common/table/components/TableBody.test.js

chalettu replied:
Went ahead and refactored the test.  Thank you for pointing me to the helpers.

sharvit commented on webpack/scenes/Settings/Tables/__tests__/TableReducer.test.js:
```diff
@@ -0,0 +1,49 @@
+} from './Table.fixtures';
+import reducer from '../TableReducer';
+
+describe('tables reducer', () => {
```
You could write much simpler test here if you would use the test-helper `testReducerSnapshotWithFixtures`.
https://github.com/Katello/katello/blob/master/webpack/move_to_pf/test-utils/testHelpers.js#L67

We still don't have an example in katello but you can see an example in foreman:
https://github.com/theforeman/foreman/blob/c28d51eb132cc079dd7a5ad7fbef63aba3cca2ab/webpack/assets/javascripts/react_app/components/PasswordStrength/__tests__/PasswordStrengthReducer.test.js

If you want to, it could be nice to use the `testReducerSnapshotWithFixtures`.

chalettu replied:
Went ahead and refactored the test.  Thank you for pointing me to the helpers.

sharvit commented on webpack/scenes/Subscriptions/SubscriptionActions.js:
```diff
@@ -76,6 +80,27 @@ export const updateQuantity = (quantities = {}) => (dispatch) => {
 };
 
+export const loadTableColumns = selectedColumns => (dispatch) => {
+  const enabledColumns = (selectedColumns ?
```
You could use

```js
export const loadTableColumns =
  (enabledColumns = SUBSCRIPTION_TABLE_DEFAULT_COLUMNS) => (dispatch) => { ... };
```

chalettu replied:
I updated this one a little bit.  Though I normally agree with your suggestion, this particular use case I ended up going with a slightly different solution.

sharvit commented on webpack/scenes/Subscriptions/SubscriptionsPage.js:
```diff
@@ -109,19 +121,40 @@ class SubscriptionsPage extends Component {
       });
   }
 
+  renderTableColumnOptions() {
```
I couldn't find where you actually use this method. Does it a leftover?

chalettu commented on webpack/move_to_pf/OptionTooltip/index.js:
```diff
@@ -0,0 +1,80 @@
+    value: PropTypes.bool,
+  })).isRequired,
+  onChange: PropTypes.func,
+  onClose: PropTypes.func.isRequired,
```
I didn't want to over engineer the component.  I could have added an onOpen method but I figured someone could add it if they felt it necessary in the future.

chalettu commented on webpack/move_to_pf/OptionTooltip/index.js:
```diff
@@ -0,0 +1,80 @@
+  renderTooltip() {
+    const { options } = this.props;
+    return (
+      <Popover id="optionTooltip" className="option-tooltip">
```
I went ahead and made it a property.  The popover bootstrap component requires an id so this is definitely one we need to include.

chalettu commented on webpack/move_to_pf/OptionTooltip/index.js:
```diff
@@ -0,0 +1,80 @@
+    value: PropTypes.bool,
+  })).isRequired,
+  onChange: PropTypes.func,
+  onClose: PropTypes.func.isRequired,
```
I went ahead and made it optional.   Good point.

chalettu commented on webpack/move_to_pf/OptionTooltip/index.js:
```diff
@@ -0,0 +1,80 @@
+    this.setState(options);
+    this.props.onChange(options);
+  }
+  renderTooltip() {
```
Popover is already a bootstrap component.  Didn't feel like needing to wrap an already written react bootstrap component.

chalettu commented on webpack/move_to_pf/OptionTooltip/index.js:
```diff
@@ -0,0 +1,80 @@
+  }
+  render() {
+    const { icon, options } = this.props;
+    const rootClose = true;
```
I went ahead and took your advice and made it a prop.

waldenraines commented:
[test katello]
